Move facility and department routes  and related type clean up

yash-learner opened the pull request:
## Proposed Changes

- Fixes second sub-task in #12909.
- Move the facility and facility-organization from central API to object specific.
- Types clean up for facility and facility-organizations and related user type clean up and a lot of random clean-ups.
- Some types clean up in role and user as they are used in facility-organization.
- Some related random clean-ups and moving something's here and there.
- Updated the useMutation to support the FormData -> https://github.com/ohcnetwork/care_fe/pull/13002/commits/e5ce12e6a9df7e24822dad617a01db964850562b

@ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [x] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Request for Peer Reviews
- [ ] Completion of QA in Mobile Devices
- [ ] Completion of QA in Desktop Devices


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added utility functions for validating latitude and longitude values.
  * Introduced new facility feature and type constants for enhanced facility categorization.
  * Added public API endpoints for retrieving facility data.

* **Refactor**
  * Updated user and facility type definitions across the app for improved consistency and specificity.
  * Replaced older user and facility types with new, more granular interfaces in forms, APIs, and components.
  * Unified and streamlined API endpoint usage for user, facility, and organization management.
  * Enhanced localization with new organization type validation messaging.

* **Bug Fixes**
  * Ensured correct handling of file uploads by properly managing request headers for FormData.

* **Chores**
  * Removed deprecated or unused facility and user types and constants.
  * Improved documentation formatting for clarity and consistency.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

yash-learner committed:
WIP: Move facilities routes

- from central api to object specific

yash-learner committed:
Move the facility cover image upload route to facility api and update the usage

yash-learner committed:
More refactor for facility types, routes, and some more

yash-learner committed:
Add separete permissions and facilitypermissions type and facilitylistread

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

yash-learner commented on src/types/facility/facilityApi.ts:
```diff
@@ -1,54 +1,73 @@
+import { FacilityCreate, FacilityListRead, FacilityRead } from "./facility";
 
 export default {
+  getAll: {
```
Move public endpoints to a separate `publicFacilityAPi` similar to schedule and patient

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/13002/commits/a1e2b46ef38cd18838b2d645179da50308bd2544

yash-learner commented on src/types/facility/facilityApi.ts:
```diff
@@ -1,54 +1,73 @@
+    method: HttpMethod.GET,
+    TRes: Type<PaginatedResponse<FacilityRead>>(),
+  },
+  getAny: {
```
Move public endpoints to a separate `publicFacilityAPi` similar to schedule and patient

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/13002/commits/a1e2b46ef38cd18838b2d645179da50308bd2544

yash-learner committed:
Move public enpoints to a separate api file, add a type for puclic facility read

and update relevant files

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

yash-learner committed:
Move the feature facility types constant to facility type and fix some type assertions

yash-learner committed:
Fix some more type assertions

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

yash-learner committed:
Update facility organization types to have common pattern for CRUD endpoints

yash-learner committed:
Move facility list users route

yash-learner committed:
Fix some type and eslint errors

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

yash-learner committed:
Clean up the userbase, userread types and remove email backend does not send that and update the usage

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

yash-learner committed:
Clean up user create and update and get

yash-learner committed:
Remove the route from api.tsx

yash-learner committed:
Clean up the old facility model file

yash-learner committed:
Fix some more type cleanup

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

yash-learner committed:
More types clean up

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

yash-learner committed:
Fix the response type

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

nihal467 commented:
check the cypress failure, its crashing in facility creation, when we click the features field

gigincg commented on src/components/Facility/FacilityForm.tsx:
```diff
@@ -25,15 +25,15 @@ import { PhoneInput } from "@/components/ui/phone-input";
-import { FacilityModel } from "@/components/Facility/models";
 
-import { FACILITY_FEATURE_TYPES, FACILITY_TYPES } from "@/common/constants";
+import { FACILITY_TYPES } from "@/common/constants";
```
Are we not moving this to facilityTypes?

yash-learner replied:
Moved https://github.com/ohcnetwork/care_fe/pull/13002/commits/ea51a8d64e1a89a2e5e754541b63f67ea56f512b 😁

gigincg commented on src/components/Facility/FacilityForm.tsx:
```diff
@@ -57,7 +57,7 @@ export default function FacilityForm({
     facility_type: z.string().min(1, t("facility_type_required")),
     name: z.string().min(1, t("name_is_required")),
-    description: z.string().optional(),
+    description: z.string().min(1, t("field_required")),
```
Is this reqiured from the backend? Why are we making it mandatory?

yash-learner replied:
Yes, it is required on backend spec https://github.com/ohcnetwork/care/blob/bd50921bdf829d1e961048c4dd2b066e8b335ef1/care/emr/resources/facility/spec.py#L33

gigincg replied:
@yash-learner An empty string is also accepted

yash-learner replied:
@gigincg Thanks for pointing out, fixed it https://github.com/ohcnetwork/care_fe/pull/13002/commits/3bdd7561fabe32eb555d36787fe7fe7d2fafa0df

yash-learner replied:
No it accepts empty strings -> reverted 

https://github.com/ohcnetwork/care_fe/pull/13002/commits/32ffef375235fa65893cd9d79c86ab5b1df77107

https://github.com/ohcnetwork/care_fe/pull/13002/commits/38241713b5845842d64e6c14f7806b52365b8f8c

gigincg commented on src/components/Facility/FacilityForm.tsx:
```diff
@@ -75,7 +75,7 @@ export default function FacilityForm({
       facility_type: "",
       name: "",
-      description: "",
+      description: undefined,
```
Not sure why this was done

yash-learner replied:
Reverted, https://github.com/ohcnetwork/care_fe/pull/13002/commits/75f45083c45adb5171bc7899846fb4a10f000573, I did it to test something

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

gigincg commented on src/components/Facility/FacilityHome.tsx:
```diff
@@ -140,12 +139,12 @@ export const FacilityHome = ({ facilityId }: Props) => {
-    const url = `${careConfig.apiUrl}/api/v1/facility/${facilityId}/cover_image/`;
+    const url = `${careConfig.apiUrl}${makeUrl(facilityApi.uploadCoverImage.path, undefined, { id: facilityId })}`;
 
     await uploadFile(
```
Can we structure this better? Or write a TODO item/file an issue for it. This feels wrong to be inside the facilityHome compoennt

yash-learner replied:
@gigincg 

Updated `query.ts` to support formdata -> https://github.com/ohcnetwork/care_fe/pull/13002/commits/e5ce12e6a9df7e24822dad617a01db964850562b#diff-d634b8361dd56c749e217ef59bf7a3a1c64ce3da3c9216cd81cc7b11ad9c8fb7

Does this look better ?

gigincg commented on src/components/Facility/FacilityHome.tsx:
```diff
@@ -335,21 +332,19 @@ export const FacilityHome = ({ facilityId }: Props) => {
-                            )}
-                        </span>
-                      </div>
+                      {facilityData?.latitude && facilityData?.longitude && (
```
Why have we discontinued using the isValidCoordinate helper?

yash-learner replied:
Because `isValidCoordinate` is incorrect and correct validation is already happening inside the `src/components/Facility/FacilityMapLink.tsx` via `isValidLatitude` and `isValidLongitude`, I just moved code related to utils.

https://github.com/ohcnetwork/care_fe/blob/e5ce12e6a9df7e24822dad617a01db964850562b/src/components/Facility/FacilityMapLink.tsx#L22

gigincg commented on src/components/Resource/ResourceForm.tsx:
```diff
@@ -50,9 +50,9 @@ import query from "@/Utils/request/query";
 import validators from "@/Utils/validators";
 import patientApi from "@/types/emr/patient/patientApi";
-import facilityApi from "@/types/facility/facilityApi";
+import PublicFacilityApi from "@/types/facility/PublicFacilityApi";
```
why haven't we named it `PublicFacilityApi` instead of `publicFacilityApi`

yash-learner replied:
It's a mistake, https://github.com/ohcnetwork/care_fe/pull/13002/commits/f18e6846a83011994a881054ef6eeac2efb723cf, fixed it

Also noticed the same mistake for `publicScheduleAppointment` will fix that in next PR, while cleaning up for the user and schedule api

gigincg commented on src/types/user/user.ts:
```diff
@@ -1,40 +1,46 @@
+  gender: (typeof GENDER_TYPES)[number]["id"];
+}
+
+export interface UserReadBase extends UserBase {
```
I think a better name would be UserReadMinimal

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/13002/commits/aa4e228627fdaa861f03ef642f09489f0693f5d4

gigincg commented on src/pages/Admin/Roles/RoleForm.tsx:
```diff
@@ -30,7 +30,7 @@ export default function RoleForm({
   const [formData, setFormData] = React.useState({
     name: role?.name || "",
```
Switch to useForm? Maybe a follow up issue can be filed

yash-learner replied:
Yep, upNext, https://github.com/ohcnetwork/care_fe/issues/13110

gigincg commented on src/pages/Facility/FacilitiesPage.tsx:
```diff
@@ -71,8 +71,11 @@ export function FacilitiesPage() {
-                setSelectedOrg(filter.organization as string);
+              if (
+                filter.organization &&
+                typeof filter.organization === "string"
```
I think we should make sure that filter.organization is a string, instead of checking using typeof

gigincg replied:
Ideally filter.organization should be string | undefined, and you can simply assign filter.organization

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/13002/commits/cd4213a659c1b22a8654812a86a3c9a7b7ba3d1e

gigincg commented on src/pages/Facility/Utils.tsx:
```diff
@@ -1,35 +1,17 @@
   if (!feature) {
     return <></>;
   }
-  const variantStyles = {
```
Why was this removed?

yash-learner replied:
Because `FACILITY_FEATURE_TYPES` already has variant as one of the field, We decided to remove it from one place.

https://github.com/ohcnetwork/care_fe/blob/cd4213a659c1b22a8654812a86a3c9a7b7ba3d1e/src/types/facility/facility.ts#L78

gigincg commented on src/pages/Facility/settings/billing/discount/discount-codes/CreateDiscountCodeSheet.tsx:
```diff
@@ -25,8 +25,8 @@ export function CreateDiscountCodeSheet() {
-    mutationFn: mutate(facilityApi.updateMonetaryComponents, {
-      pathParams: { facilityId },
+    mutationFn: mutate(facilityApi.setMonetaryComponents, {
+      pathParams: { id: facilityId },
```
This would've been better left as facilityId, and the pathparam also left as facilityId

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/13002/commits/d5ad02f14dc957546b517489774e5bb0540e2539

gigincg commented on src/pages/Facility/settings/billing/settings/BillingSettings.tsx:
```diff
@@ -28,7 +28,7 @@ export function BillingSettings() {
   } = useMutation({
-    mutationFn: mutate(facilityApi.patchInvoiceNumberExpression, {
+    mutationFn: mutate(facilityApi.setInvoiceExpression, {
       pathParams: { id: facilityId },
```
Same here

yash-learner replied:
https://github.com/ohcnetwork/care_fe/commit/d5ad02f14dc957546b517489774e5bb0540e2539

gigincg commented on src/pages/Facility/settings/organizations/components/EditFacilityUserRoleSheet.tsx:
```diff
@@ -69,7 +69,7 @@ export default function EditUserRoleSheet({
 
   const { mutate: updateRole } = useMutation({
-    mutationFn: (body: { user: string; role: string }) =>
+    mutationFn: (body: { role: string }) =>
```
I don't think you need to use a function here at all, you can just specifuy the mutationFn as `mutate(api, {pathParams})`

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/13002/commits/191aa70d81c184bb5113bd544792b9a91ad0970b

gigincg commented on src/pages/Facility/settings/organizations/components/EditFacilityUserRoleSheet.tsx:
```diff
@@ -153,9 +152,6 @@ export default function EditUserRoleSheet({
                 <span className="font-medium text-lg">
                   {formatName(userRole.user)}
                 </span>
-                <span className="text-sm text-gray-500">
```
Was this always being printed as undefined? Why was this removed?

yash-learner replied:
Backend is not sending email field for this endpoint, look like till now it survived without any error because of being present in our Frontend Types

Right now, BE sends email for only one endpoint

gigincg commented on src/pages/Facility/settings/organizations/components/FacilityOrganizationFormSheet.tsx:
```diff
@@ -81,8 +80,9 @@ export default function FacilityOrganizationFormSheet({
-    description: z.string().optional(),
-    org_type: z.enum(["dept", "team"]),
+    description: z.string().trim(),
+    org_type: z.enum(["dept", "team"] as const),
```
should ideally be an enum in the typefile for facilityOrganization

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/13002/commits/78b861140cd9b69b23c50dba363d1b01d20f35bf

yash-learner committed:
Incldue ts files in tree shaking and remove old res

gigincg commented on src/pages/Facility/settings/organizations/components/FacilityOrganizationFormSheet.tsx:
```diff
@@ -127,7 +129,7 @@ export default function FacilityOrganizationFormSheet({
 
   const { mutate: updateOrganization, isPending: isUpdating } = useMutation({
-    mutationFn: (body: FacilityOrganizationEdit) =>
+    mutationFn: (body: FacilityOrganizationCreate) =>
```
Remove the function here

gigincg commented on src/pages/Facility/settings/organizations/components/FacilityOrganizationFormSheet.tsx:
```diff
@@ -144,15 +146,16 @@ export default function FacilityOrganizationFormSheet({
 
     if (isEditMode) {
       updateOrganization({ ...data, facility: facilityId });
     } else {
```
Why was this changed? Was the API not working? Or was the API changed?

yash-learner replied:
> Why was this changed?

This was added because backend write spec expects that field, and BE sets to True (default value).

https://github.com/ohcnetwork/care/blob/c4ad067f0b13cf04a9d94b0606798bdb3c6c50ac/care/emr/resources/facility_organization/spec.py#L24

Added it match BE, 

What do you think ?

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/13002/commits/3c0805ded11820d3beb57b36ee18ddc4832519ae

gigincg commented on src/pages/PublicAppointments/Success.tsx:
```diff
@@ -21,7 +21,9 @@ export function AppointmentSuccess(props: { appointmentId: string }) {
   const tokenData = patientUserContext?.tokenData;
 
-  const userData: UserBase = JSON.parse(localStorage.getItem("user") ?? "{}");
+  const userData: UserReadBase = JSON.parse(
```
Move to a helper maybe?

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/13002/commits/225c8b44408ba9a6d7ce05aa1cedba37ae14bb62

gigincg commented on src/types/emr/permission/permission.ts:
```diff
@@ -4,3 +4,12 @@ export interface Permission {
+}
+
+export interface FacilityPermissions extends Permissions {
+  root_org_permissions: string[];
```
In order to cleanly manage all permission based rendering, this should ideally be an enum of text values, one for each backend permission.

Making it a follow up issue is also fine

yash-learner replied:
https://github.com/ohcnetwork/care_fe/issues/13131

gigincg commented on src/types/facility/PublicFacilityApi.ts:
Rename to publicFacilityApi.ts

gigincg commented on src/types/facility/facility.ts:
```diff
@@ -8,40 +15,39 @@ export interface FacilityBareMinimum {
+  read_cover_image_url?: string;
+  cover_image_url?: string;
+  geo_organization: Organization;
+  latitude?: string;
```
Valid inconsistency, lat long needs to be either string/number

yash-learner replied:
Fixed https://github.com/ohcnetwork/care_fe/pull/13002/commits/117fd9c5b8fa942390ea35ffee83c05c1e7accbf#diff-c2b540bed3cbc501f5f9ce1d1dfe6e688b55c1ef3ec491446c6c209d4c632488

yash-learner committed:
Fix the params in resource form

yash-learner commented:
@nihal467 

> check the cypress failure, its crashing in facility creation, when we click the features field


>  its crashing in facility creation

It is happening because of tree shaking that we do in the deployment's, now I had added look for icons in the `ts` files as well, fixed it.

https://github.com/ohcnetwork/care_fe/pull/13002/commits/9c2aa240c76568ddfc5befe2c875ce84406511ef

fixed the cypress as well https://github.com/ohcnetwork/care_fe/pull/13002/checks

yash-learner committed:
Move facility types to facility type file from constants

yash-learner committed:
Keep default value as empty string, instead of undefined

yash-learner committed:
WIP: Update query.ts to support form data

yash-learner committed:
Fix the naming mistake for public facility api

yash-learner committed:
Rename UserReadBase to UserReadMinimal

yash-learner commented on plugins/treeShakeCareIcons.ts:
```diff
@@ -48,7 +48,7 @@ export function treeShakeCareIcons(
   // Finds all used icon names within the project's source files (`.tsx` or `.res` extensions).
   function getAllUsedIconNames() {
-    const files = globSync(path.resolve(rootDir, "{apps,src}/**/*.{tsx,res}"));
+    const files = globSync(path.resolve(rootDir, "{apps,src}/**/*.{tsx,ts}"));
```
We no longer have `.res` files in codebase

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

yash-learner committed:
Revert the typeof check

yash-learner committed:
Call it facilityId

yash-learner committed:
Some cleanup related to FacilityOrganization type

yash-learner committed:
Remove explicit body and some type cleanup

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

yash-learner committed:
Fix pulled changes

yash-learner committed:
Move the deo relaetd to set and get of user from localstorage to type file

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

yash-learner committed:
Keep types of lat & long consistent

yash-learner commented on src/types/facility/facility.ts:
```diff
@@ -8,40 +15,39 @@ export interface FacilityBareMinimum {
-  root_org_permissions: string[];
-  child_org_permissions: string[];
+  latitude?: string;
+  longitude?: string;
```
@cursor fixed it https://github.com/ohcnetwork/care_fe/pull/13002/commits/117fd9c5b8fa942390ea35ffee83c05c1e7accbf#diff-c2b540bed3cbc501f5f9ce1d1dfe6e688b55c1ef3ec491446c6c209d4c632488

yash-learner commented on src/components/Users/UserForm.tsx:
```diff
@@ -70,11 +70,14 @@ export default function UserForm({
+        "staff",
+        "volunteer",
+        "administrator",
+      ]),
```
Because backend spec has it as required

yash-learner committed:
Revert the active being passed

as backend always defaults to true and currently we are not allowing editing of active from FE

yash-learner commented on src/components/Facility/FacilityForm.tsx:
```diff
@@ -185,18 +192,14 @@ export default function FacilityForm({
-        geo_organization: (
-          facilityData.geo_organization as unknown as Organization
-        )?.id,
+        geo_organization: facilityData.geo_organization.id,
```
@cursor you are wrong

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

yash-learner committed:
Do not allow the empty strings

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

nihal467 commented:
<img width="1893" height="763" alt="image" src="https://github.com/user-attachments/assets/0d57bed6-0324-4242-8427-6a127e5ed780" />

- unable to remove existing suffix in the user profile

yash-learner committed:
Revert description being mandatory for facility

BE accepts empty string but not null -> so deault to "" in zod and just string in type

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

yash-learner committed:
Default to "" for FO desc as well

and add trim

yash-learner committed:
Fix the suffix and prefix not being cleared

nihal467 commented:
LGTM

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

yash-learner committed:
Merge branch 'develop' into issues/12909/migrate-from-cr-osr-pr2

yash-learner committed:
Merge remote-tracking branch 'upstream/develop' into issues/12909/migrate-from-cr-osr-pr2

AdityaJ2305 commented on src/pages/Facility/settings/organizations/components/FacilityOrganizationFormSheet.tsx:
```diff
@@ -81,16 +73,26 @@ export default function FacilityOrganizationFormSheet({
-    org_type: z.enum(["dept", "team"]),
+    description: z.string().trim().default(""),
+    org_type: z.nativeEnum(FacilityOrganizationType).refine(
+      (val) => {
+        return (
+          val === FacilityOrganizationType.DEPT ||
+          val === FacilityOrganizationType.TEAM
+        );
+      },
+      {
```
why ?? 

@yash-learner

yash-learner replied:
@AdityaJ2305 

Because we allow only selecting the dept and team types but not root so

and

I guess even backend does not allow that

AdityaJ2305 replied:
Yes , but I don't know why it looks off to me could you 100% personal choice 

but we are already doing these below things to stop selecting `root` 

- Filtering select content to show only dept and team (that is more enough I think )
- we are not showing edit for root org 
- In case if in future we do enable editing for root org , it would fail to pass 

<img width="815" height="466" alt="Screenshot 2025-08-14 at 9 56 10 PM" src="https://github.com/user-attachments/assets/447ab16d-e521-40a4-b245-704e2fdb44a4" />

yash-learner replied:
>  but I don't know why it looks off to me 

Why ? @AdityaJ2305 

---

> Filtering select content to show only dept and team (that is more enough I think )

Yes

but zod schema field should match expected values right otherwise it is like faking right how we type cast ?, we allow only `Team` and `Dept` right in select field, → What do you think ?

The reason I did refine and did validation is that without that we need to create a separate type for this case in the form for orgType and API always returns all three `root`, `team`. `dept`  this would cause type mismatch and this would need to handled two different types and needs type casting when handling the data from API with form when editing which is definitely not good


> In case if in future we do enable editing for root org , it would fail to pass

Think about now, we are not allowing, right ?

Right now, root is created as system generated 

```py
           facility_organization = FacilityOrganization.objects.create(
                org_type="root",
                name="Administration",
                system_generated=True,
                facility=self,
            )
 ```

Why would you allow editing a system generated one by user/client ?

Definitely we will not allow AFAIK

In case we allow it then we will update it is really a tiny change and to allow that you need to make change to FE select drop down list first then you can remove this refine in zod schema 🤷‍♂️

AdityaJ2305 replied:
ok 👍

yash-learner replied:
@AdityaJ2305 

https://github.com/ohcnetwork/care/pull/3195/files
